### DISCUSSION TITLE
[BUG?] Cannot create Noticed::Event in MariaDB

### DISCUSSION THREAD

**Discussion by janpeterka** [excid3/noticed#412] [Q&A]
## Bug Report
Maybe this is wrong place to report this - I'm not sure if there's something wrong in this gem, or it's problem with ActiveSupport or some misconfiguration on my side.
Will be glad for help with that though, and if it seems like this is not `noticed`-related, I will understand if this will be just closed.

**Describe the Bug:**
Running tests locally, I have no problem. Running them on my Gitlab runner, I'm getting error:
```
     Failure/Error:
       NewAnnouncementNotifier.with(record: @announcement, message: 'nový příspěvek')
                              .deliver(@announcement.group.members - [Current.member])
     ActiveRecord::StatementInvalid:
       Mysql2::Error: CONSTRAINT `params` failed for `sborio_tests`.`noticed_events`
     # ./app/controllers/announcements_controller.rb:23:in `create'
     # ./spec/requests/announcements_controller_spec.rb:18:in `block (3 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Mysql2::Error:
     #   CONSTRAINT `params` failed for `sborio_tests`.`noticed_events`
     #   ./app/controllers/announcements_controller.rb:23:in `create'
```

Digging deeper, I found out the difference is that my local tests are running on MySQL 8.0, and on runner I'm using MariaDB 10.11.6, which behave differently, using LONGTEXT and `json_valid` constaint.

I then tried to get more info from console:
```bash
irb(main):004> NewAnnouncementNotifier.with(record: Announcement.first, message: 'nový příspěvek').save!
  Announcement Load (0.4ms)  SELECT `information`.* FROM `information` ORDER BY `information`.`id` ASC LIMIT 1
  TRANSACTION (0.2ms)  BEGIN
  NewAnnouncementNotifier Create (0.7ms)  INSERT INTO `noticed_events` (`type`, `record_type`, `record_id`, `params`, `created_at`, `updated_at`, `notifications_count`) VALUES ('NewAnnouncementNotifier', NULL, NULL, '{\"message\"=>\"nový příspěvek\", \"_aj_symbol_keys\"=>[\"message\"]}', '2024-02-15 11:02:33.460483', '2024-02-15 11:02:33.460483', NULL)
  TRANSACTION (0.2ms)  ROLLBACK
/home/gitlab-runner/builds/cg3oSeWc/0/ninjas/sborio-on-rails/vendor/bundle/ruby/3.2.0/gems/mysql2-0.5.6/lib/mysql2/client.rb:151:in `_query': Mysql2::Error: CONSTRAINT `params` failed for `sborio_tests`.`noticed_events` (ActiveRecord::StatementInvalid)
/home/gitlab-runner/builds/cg3oSeWc/0/ninjas/sborio-on-rails/vendor/bundle/ruby/3.2.0/gems/mysql2-0.5.6/lib/mysql2/client.rb:151:in `_query': CONSTRAINT `params` failed for `sborio_tests`.`noticed_events` (Mysql2::Error)
```

I guess it has trouble with having `=>` instead of `:` in the saved hash.
Why do we get that? [Coder](https://github.com/excid3/noticed/blob/main/lib/noticed/coder.rb#L4) (or used `ActiveJob::Argument`, to be precise) does it, if I understand that code correctly.

But! When I run the same code in local environment, I'm getting correct json:
```
irb(main):001> NewAnnouncementNotifier.with(record: Announcement.first, message: 'nový příspěvek').save!
  Announcement Load (0.7ms)  SELECT `information`.* FROM `information` ORDER BY `information`.`id` ASC LIMIT 1
  TRANSACTION (0.5ms)  BEGIN
  NewAnnouncementNotifier Create (1.8ms)  INSERT INTO `noticed_events` (`type`, `record_type`, `record_id`, `params`, `created_at`, `updated_at`, `notifications_count`) VALUES ('NewAnnouncementNotifier', NULL, NULL, '{\"message\":\"nový příspěvek\",\"_aj_symbol_keys\":[\"message\"]}', '2024-02-15 11:33:07.414884', '2024-02-15 11:33:07.414884', NULL)
  TRANSACTION (3.2ms)  COMMIT
=> true
```

So I'm really just confused on what's happening here :sweat_smile: 

Will be extremely grateful for any guidance!

**Environment:**
- Noticed gem version: 2.1.3
- Ruby version: 3.2.2
- Rails version: 7.1.3
- Operating System: Ubuntu 23.10

**Checklist:**

- [ x] I have searched for similar issues and couldn't find any
- [ x] I have checked the documentation for relevant information
- [ x] I have included all the required information

**Comment by excid3** [comment]
MariaDB JSON columns are actually text columns with JSON constraints, so they don't work interchangeably with MySQL.

I don't have any plans to support MariaDB but if someone wants to make a PR we can.

**Comment by jronallo** [comment]
I ran into this issue with MariaDB too and was able to work around it by modifying the migration to use type longtext for params and overriding the Event class to include this line:

```ruby
serialize :params, JSON
```

Not tested significantly, but so far it appears to be working.
